Sources/Foundation/NSString.swift

```diff
@@ -1359,7 +1359,7 @@ extension NSString {
         // just copy for now since the internal storage will be a copy anyhow
         self.init(bytes: bytes, length: len, encoding: encoding)
         if freeBuffer { // don't take the hint
-            free(bytes)
+          bytes.deallocate()
```

Comment:
Hmm, this is documented to call `free`: https://developer.apple.com/documentation/foundation/nsstring/1413830-initwithbytesnocopy

I think we should replicate this comment and the caller should make sure that the memory is allocated using `malloc`. WDYT?

Reply:
So:
 - On Linux and Darwin, `free()` is ultimately compatible with the allocation done by CF, ObjC and Swift. This is sort of a non-issue on those platforms.
- On other platforms, there _can_ be a distinction.

This has two sides: a contract side and an implementation side.

On the contract side: in ObjC, the original API here takes a `void *` and the language does _not_ mandate how you produce that pointer, only that it be valid. It makes sense here to specify precisely the deallocation strategy.

_But._ This is _Swift_. And in Swift, this type is `UnsafeMutableRawPointer` — a type that _does_ have a language-provided default deallocation facility. And in this language, since there are far fewer guarantees wrt the memory layout of language constructs*, the vast majority of these pointers will be produced by the `.allocate(byteCount:alignment:)` method of that class or the equivalent bound variations in the non-`Raw` pointers.

My opinion here is that we can deviate from Darwin Foundation as follows:

 - `.deallocate()` will eventually call through to `free()` on Linux and Darwin. We are going to rely on this property, for now, or conditionality the change away from `free()` on these OSes, so long that it interoperates correctly with buffers allocated with `.allocate(…)`.

- On Windows and other OSes where it's not guaranteed that `.allocate(…)` use malloc, we mandate that `freeWhenDone: true` use `.deallocate()` instead of `free()`. This puts an onus on the caller to use `.allocate(…)` rather than use OS facilities directly. I think this is a reasonable requirement to make: callers that choose a nonstandard allocation strategy on other OSes are already mandated to use `freeWhenDone: false` with these pointers and handle deallocation themselves; `malloc()` just happens to not be the standard on those OSes.

The contract side is straightforward-ish — we're just extending the existing contract to go from 'specifically malloc()' to 'the system-preferred deallocation strategy, which on Linux and Darwin is malloc()'. The implementation side puts a burden on implementors on those OSes where this is not the case — which means that existing setups don't regress.

I think I am in favor of making this change conditional on the OS, and we should note it in the release notes (in Docs/ReleaseNotes*.md) as a well-known deviation from Darwin behavior.

* that is, you "can't just" take a pointer to a Swift.Array like you can for an array in C, and have strong guarantees about the memory layout of that object unless you are working hand in hand with the compiler

Reply:
@millenomi but what happens to all the code that was previously correctly matching `malloc` with `freeWhenDone: true`? That works today and will then break, correct?

Reply:
@weissi I actually have code that passes raw buffer originated from C module (which uses malloc internally) to `String(bytesNoCopy:...` with "freeWenDone: true". And it crashes with this patch. I understand motivation though.

@compnerd I think we need to add one more detail to the docs. Official documentation doesn't clarify _when_ `freeWhenDone` takes effect. It is unclear that on Windows we are allowed to do manual cleanup right after String init. As user, I'd rather think that `String(bytesNoCopy:..` will need specified buffer alive until deinit.

Reply:
@lxbndr Thanks for chiming in! I mean it is documented to `free` so I think it should `free`. If we think we need a Swift version of this, why not create a new `String(bytesNoCopy:...deallocateWhenDone:)` that would call deallocate?

I'm pretty strongly against just changing the contract.

Reply:
@lxbndr Are you saying that memory allocated with `malloc` but then passed as an `UnsafeRawMutableBufferPointer` then has `deallocate()` called on it crashes? Is this on windows? Because that sounds like another bug. 

@weichsel `Data` has the following initialiser: `init(bytesNoCopy bytes: UnsafeMutableRawPointer, count: Int, deallocator: Data.Deallocator)` where the deallocator can be specified including a custom one passed with a closure so that probably offers a good way to do it although it needs `deallocate()` added into `Data.Deallocator` https://developer.apple.com/documentation/foundation/data/deallocator

Reply:
@spevans Yes, this is on Windows. deallocateWhenDone:false together with manual free() works like a charm. Initial PR description now seems about a bit different case. It's about Swift buffers which can't be passed to free(). I think I need look closer on my case.

Reply:
@weissi Feels like band-aid, I agree. Approach with custom deallocator like in `Data` looks way better.

Reply:
@lxbndr We're looking into a more comprehensive solution that will fix the issue for good, but a quick search on GitHub shows that using `UMRP.alloc…(…)` with `freeWhenDone:` is ubiquitous and prevalent.